Make GenericController get data from the store directly.

tomalec opened the pull request:
:warning: This PR requires https://github.com/woocommerce/woocommerce/pull/50746/

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->
This is a more gracefully step towards the idea from https://github.com/woocommerce/woocommerce/pull/50747

This allows the creation of new analytics reports without the need to create the Query layer. So, extension developers could use Controller to handle default query args in its schema and get data directly from the data store.

Changes the `GenericController` API not to suggest the use of `Query` so we could potentially refactor existing analytic controllers to remove Queries as well.

I'd rather not add dedicated changelog here, this PR adds to changes already  merged in the feature branch

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

This is a developer/code-oriented change. It does not change the functionality.

1. Review the code. Check for its readability. And if the suggested example is actually simpler and easier to implement.
2. Run automated tests
3. Smoke test for regression

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>
…-internal

tomalec committed:
Make `GenericController` get data from the store directly.

mikkamp commented on plugins/woocommerce/src/Admin/API/Reports/GenericController.php:
```diff
@@ -41,11 +41,10 @@
+ * (To change this behavior, override the `get_datastore_data()` method).
  *
- * To use it, please register your controller in the `woocommerce_admin_rest_controllers`
+ * To use the controller, please register your it in the `woocommerce_admin_rest_controllers`
```
```suggestion
 * To use the controller, please register it with the filter `woocommerce_admin_rest_controllers`
```

mikkamp commented on plugins/woocommerce/src/Admin/API/Reports/DataStore.php:
```diff
@@ -48,9 +48,16 @@
  * or using {@link \WC_Data_Store::load() \WC_Data_Store::load()}.
  *
+ * We recommend registering using the REST base name of your Controller as the key, e.g.:
+ * <pre><code class="language-php">add_filter( 'woocommerce_data_stores', function( $stores ) {
```
Are these HTML tags specifically being added for generated documentation? Doesn't seem to be a convention for any of the other code within WC.

tomalec replied:
Yes, this is the way to generate a code block in [code-reference](https://woocommerce.github.io/code-reference/)

https://github.com/woocommerce/code-reference/pull/25

> Doesn't seem to be a convention for any of the other code within WC.

I found that some places use `@example` to point to a `.md` file. But I think it does not serve our readers well, as code-reference leaves the file name as plain text, not even a link.

mikkamp commented on plugins/woocommerce/src/Admin/API/Reports/GenericController.php:
```diff
@@ -99,16 +98,14 @@ public function add_pagination_headers( $request, $response, int $total, int $pa
-	protected function construct_query( $query_args ) {
-		return new GenericQuery( $query_args, $this->rest_base );
+	protected function get_datastore_data( $query_args = array() ) {
+		$data_store = \WC_Data_Store::load( $this->rest_base );
```
Do we need to clarify somewhere here that this will throw an exception, if the datastore can't be loaded? I know we clarify in the DataStore class that it should match the rest base name, but if we get the error we'll be looking at this class.

Especially since it seems to be an uncaught exception so not handled gracefully when this is wrong.

tomalec replied:
:thinking: It's hard to catch it as it's just `Exception` as any other.

Do you think a code-doc is enough? https://github.com/woocommerce/woocommerce/pull/50749/commits/d4ccb27c05cba5b373fd1e5c584b76fbac33e06e

mikkamp replied:
I think the code doc should be enough. After all it's a developer error that they didn't use the class correctly. So we don't necessarily need it to be graceful for the end consumer.

mikkamp commented on plugins/woocommerce/src/Admin/API/Reports/GenericController.php:
```diff
@@ -99,16 +98,14 @@ public function add_pagination_headers( $request, $response, int $total, int $pa
 	/**
-	 * Forwards a Query constructor,
-	 * to be able to customize Query class for a specific report.
+	 * Get data from `report-{$this->rest_base}` store, based on the given query vars.
```
As far as I can see we are now no longer adding the prefix `report-`, is that correct?

Also `$this->rest_base` seems to be something like `reports/customers`, whereas the datastore should be loaded with `report-customers`. Notice the missing `s`. I tested this by removing one of the `get_datastore_data` functions and it doesn't work correctly unless I pass a different variant of `rest_base`.

tomalec replied:
> As far as I can see we are now no longer adding the prefix report-, is that correct?

Correct.

> Also $this->rest_base seems to be something like reports/customers, whereas the datastore should be loaded with report-customers. Notice the missing s. I tested this by removing one of the get_datastore_data functions and it doesn't work correctly unless I pass a different variant of rest_base.

That's why [we ask devs to ](https://github.com/woocommerce/woocommerce/pull/50749/files#diff-a645b0e958454758133803c533bd522179b0cacd086047ef5d8a5d320f1ec42bR54)
> ```
> registering using the REST base name of your Controller as the key, e.g.:
>  * <pre><code class="language-php">add_filter( 'woocommerce_data_stores', function( $stores ) {
>  * $stores['reports/my-thing'] = 'MyExtension\Admin\Analytics\Rest_API\MyDataStore';
> ```

So, to use the `reports/customers` as the identifier, not the `report-customers`. At the end of the day this is the Resource Identifier

I made an example of queryless Taxes DatsStore at https://github.com/woocommerce/woocommerce/compare/dev/queryless-datastores...dev/queryless-datastores-taxes?expand=1

Also, see the discussion at pfKYZu-Le-p2#comment-653 & pfKYZu-Le-p2#comment-64.  Where I suggest using one, single resource identifier for REST URI, DataStore key, context, and cache key.

mikkamp replied:
Ok, that makes sense. As long as it's clear in the examples we provide then we shouldn't get developers running into issues.

Thanks for the additional discussion links, that wasn't clear from looking just at the PR.

mikkamp reviewed (commented):
Thanks for the changes, it is working as expected for the existing controllers.

I noticed that for any controller that doesn't override the `get_datasore_data` function and make use of the `GenericQuery` class it will no longer filter the query args + data when fetching it from the datastore.

Is that intentional? I saw in the original PR you mentioned:
> Move filters call to controller or datastore

It's not clear if that is to be done when the Query class gets removed further or at what stage.

I also added some additional comments as I went through testing the code.

tomalec committed:
Update plugins/woocommerce/src/Admin/API/Reports/GenericController.php

Co-authored-by: Mik <mikkamp@users.noreply.github.com>

tomalec commented:
Thanks for the review :) 

> I noticed that for any controller that doesn't override the get_datasore_data function and make use of the GenericQuery class it will no longer filter the query args + data when fetching it from the datastore.
> 
> Is that intentional? I saw in the original PR you mentioned:
> 
> Move filters call to controller or datastore
> 
> It's not clear if that is to be done when the Query class gets removed further or at what stage.


Current, Core analytics use Query classes, and use those filters there. Personally, I think the better place for `woocommerce_analytics_*_select_query`, `woocommerce_analytics_*_query_args` filter calls would be the `DataStore::get_data` as that is the place that ultimately process the query args and makes the select query.
In the ideal world, implementing it from scratch, I'd add it there.

However, changing that flow for existing classes could be too much of a breaking change. That's why we cannot add it to the shared `DataStore` class, unfortunately. (Unless, we agree to do a breaking change or make a separate DataStore class for the new implementations )

That's why I was thinking this way: those filters are here to make WooCommerce Core analytics extensible. If an extension developer would like to make their custom analytics extensible, it's up to them to define their own filters. So we could suggest in guidelines, that they may consider adding such filters in their own `DataStore::get_data`, like:

```php
class MyDataStore extends ReportsDataStore implements DataStoreInterface {     
	public function get_data( $query_args ) {	
		$filtered_args       = apply_filters( 'myextension_analytics_unicorns_query_args', $query_args );
		$results    = parent::get_data( $filtered_args );
		return apply_filters( 'myextension_analytics_unicorns_select_query', $results, $args );
    }
}
```

tomalec committed:
Add `FilteredGetDataTrait` to simplify filtering queryless datastores.

tomalec committed:
Add code doc about thrown error to `get_datastore_data`

Address https://github.com/woocommerce/woocommerce/pull/50749#discussion_r1721899560

tomalec committed:
Fix store name in codedoc

Address https://github.com/woocommerce/woocommerce/pull/50749#discussion_r1721918176

tomalec commented:
Thank you, @mikkamp, for the detailed review. I answered your comments, would you mind another look?

tomalec committed:
Merge branch 'dev/analytics-code-docs' into dev/queryless-datastores

mikkamp commented on plugins/woocommerce/src/Admin/API/Reports/FilteredGetDataTrait.php:
```diff
@@ -0,0 +1,59 @@
+		 */
+		return apply_filters( "woocommerce_analytics_{$this->context}_select_query", $results, $args );
+	}
+
```
This extra empty line is failing in the linter:
> ```59 | ERROR | The closing brace for the trait must go on the next line after the body (PSR2.Classes.ClassDeclaration.CloseBraceAfterBody)```

mikkamp commented on plugins/woocommerce/src/Admin/API/Reports/FilteredGetDataTrait.php:
```diff
@@ -0,0 +1,59 @@
+		/**
+		 * Called before the data is fetched.
+		 *
+		 * @since x.x.x
```
This is also failing in the linter:
> ```47 | ERROR | A "@since" tag was found but no version declared. Required format <x.x> or <x.x.x> (WooCommerce.Commenting.CommentHooks.MissingSinceVersionComment)```

I can't find documentation on how this since tag should be populated, but in [other PR's](https://github.com/woocommerce/woocommerce/pull/48295/files#diff-9f485fafffab0cb09112e052de4a942574cdf23b96d034eb140d39c308f1e05cR95) I'm seeing a specific version being put.

mikkamp replied:
After discussion, this is going to be fixed in the feature branch when we know for certain which version we will target. So this can be ignored.

tomalec replied:
We will fix those in the feature branch, for all the `x.x.x` together to make sure to put the correct version everywhere.

mikkamp commented on plugins/woocommerce/src/Admin/API/Reports/FilteredGetDataTrait.php:
```diff
@@ -0,0 +1,59 @@
+		 * Called after the data is fetched.
+		 * The results can be modified here.
+		 *
+		 * @since x.x.x
```
Same as previous comment about this linter failure.

mikkamp replied:
Same as previous comment, we can ignore.

mikkamp reviewed (commented):
Thanks for the additional changes and clarifications.

The E2E tests all passed but the [linter is failing](https://github.com/woocommerce/woocommerce/actions/runs/10461282853/job/28969303470?pr=50749). I just left a few comments as I think we should be able to fix those before merging.

mikkamp reviewed (approved):
I'll go ahead and approve this PR so it can get merged to the feature branch.

There is just this one linter error missing to address: https://github.com/woocommerce/woocommerce/pull/50749#discussion_r1723175264

tomalec committed:
Fix whitespace for PHPCS
